crates/rspack_plugin_circular_dependencies/src/lib.rs

```diff
@@ -0,0 +1,385 @@
+#![feature(array_windows)]
+
+use std::sync::Arc;
+
+use derive_more::Debug;
+use itertools::Itertools;
+use rspack_collections::{Identifier, IdentifierMap, IdentifierSet};
+use rspack_core::{
+  ApplyContext, Compilation, CompilerOptions, ModuleIdentifier, Plugin, PluginContext,
+};
+use rspack_core::{CompilationOptimizeModules, DependencyType};
+use rspack_error::{Diagnostic, Result};
+use rspack_hook::{plugin, plugin_hook};
+use rspack_regex::RspackRegex;
+
+struct CycleDetector<'a> {
+  module_map: &'a IdentifierMap<GraphModule>,
+}
+
+impl<'a> CycleDetector<'a> {
+  fn new(module_map: &'a IdentifierMap<GraphModule>) -> Self {
+    Self { module_map }
+  }
+
+  fn get_module(&self, id: &ModuleIdentifier) -> &GraphModule {
+    self
+      .module_map
+      .get(id)
+      .expect("Module map should only contain references to existing modules")
+  }
+
+  /// Returns all dependency cycles contained in the dependency graph starting at
+  /// `initial_module_id`.
+  fn find_cycles_from(
+    &mut self,
+    initial_module_id: ModuleIdentifier,
+  ) -> Vec<Vec<ModuleIdentifier>> {
+    let mut cycles = vec![];
+    self.recurse_dependencies(
+      initial_module_id,
+      &mut IdentifierSet::default(),
```

Comment:
Should we use `&mut Identifier::default()` here ?

This is used for reduce dup visit I suppose, should we move this up to the upper call site ?

Assume entry module a and entry module b both depend on module shared, shared is the root of a cycle, if we make seen_set a new HashSet everytime we will end up 2 same cycle

Reply:
I think re-reporting each cycle for each entrypoint is _somewhat_ useful, purely for the sake of completeness (i.e., you can know that a cycle is encountered in 2 of 5 entrypoints, but not in the other 3, and use the `onDetected` hook to report them separately). I don't feel super strongly about that, though.